File: 202. Happy Number/solution.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
  let slow = (fast = n);
  do {
    slow = digitSquareSum(slow);
    fast = digitSquareSum(fast);
    fast = digitSquareSum(fast);
  } while (slow !== fast);
  return slow === 1;
};

function digitSquareSum(n) {
  let sum = 0;
  while (n) {
    sum += Math.pow(n % 10, 2);
    n = Math.floor(n / 10);
  }
  return sum;
}

/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
  let sum = 0;
  let exist_nums = new Set();
  while (sum !== 1) {
    n = n
      .toString()
      .split("")
      .reduce((p, c) => {
        return p + c * c;
      }, 0);
    sum = n;
    if (exist_nums.has(sum)) {
      return false;
    }
    exist_nums.add(sum);
  }
  return true;
};

File: 202. Happy Number/solution.go
func isHappy(n int) bool {
	appearNum := make(map[int]bool)

	for n > 1 {
		newN := getDigitSquareSum(n)
		if appearNum[newN] {
			return false
		}
		appearNum[newN] = true
		n = getDigitSquareSum(newN)
	}
	if n == 1 {
		return true
	}
	return false
}
func getDigitSquareSum(n int) int {
	sum := 0
	for n > 0 {
		sum += (n - (n/10)*10) * (n - (n/10)*10)
		n = n / 10
	}

	return sum
}